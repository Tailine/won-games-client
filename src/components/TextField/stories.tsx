import { Story, Meta } from '@storybook/react/types-6-0'
import TextField, { TextFieldProps } from '.'
import { MailOutline } from '@styled-icons/material-outlined/MailOutline'

export default {
  title: 'Form/TextField',
  component: TextField,
  args: {
    label: 'E-mail',
    labelFor: 'Email',
    id: 'Email',
    initialValue: '',
    placeholder: 'john.cage@gmail.com'
  },
  argTypes: {
    onInput: { action: 'changed' }
  }
} as Meta

export const Default: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

Default.args = {
  icon: <MailOutline />
}

export const WithError: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

WithError.args = {
  error: 'Ops...something went wrong'
}
