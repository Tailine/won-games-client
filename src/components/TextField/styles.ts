import styled, { css } from 'styled-components'
import { DefaultTheme } from 'styled-components'
import { TextFieldProps } from '.'

type IconPositionProps = Pick<TextFieldProps, 'iconPosition'>

export const InputWrapper = styled.div<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};
    align-items: center;
    flex-direction: ${iconPosition === 'right' ? 'row-reverse' : 'row'};
    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    svg {
      width: 2.2rem;
      fill: ${theme.colors.gray};
    }
  `}
`

export const Input = styled.input<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
  color: ${theme.colors.black};
  font-family: ${theme.font.family};
  font-size: ${theme.font.sizes.medium};
  padding: ${theme.spacings.xxsmall} 0;
  padding-${iconPosition}: ${theme.spacings.xxsmall};
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;

    &::placeholder {
      color: ${theme.colors.gray};
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
  `}
`

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Input} {
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentColor;
      }
    }
  `,
  error: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border: 1px solid ${theme.colors.error};
    }
    ${Label} {
      color: ${theme.colors.error};
    }
    svg {
      fill: ${theme.colors.error};
    }
  `
}

type WrapperProps = {
  error: boolean
} & Pick<TextFieldProps, 'disabled'>

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled, error }) => css`
    ${disabled && wrapperModifiers.disabled(theme)}
    ${error && wrapperModifiers.error(theme)}
  `};
`

export const ErrorMessage = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.error};
    font-size: ${theme.font.sizes.xsmall};
  `};
`
