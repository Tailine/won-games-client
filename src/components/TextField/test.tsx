import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { MailOutline } from '@styled-icons/material-outlined/MailOutline'

import { renderWithTheme } from 'utils/tests/helpers'

import TextField from '.'

describe('<TextField />', () => {
  it('renders with Label', () => {
    renderWithTheme(<TextField label="Label" labelFor="Field" id="Field" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('renders without Label', () => {
    renderWithTheme(<TextField />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('renders with placeholder', () => {
    renderWithTheme(<TextField placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('changes its value when typing', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        onInput={onInput}
        label="TextField"
        labelFor="TextField"
        id="TextField"
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })
    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('is accessible by tab', async () => {
    renderWithTheme(
      <TextField label="TextField" labelFor="TextField" id="TextField" />
    )

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()

    await waitFor(() => {
      expect(input).toHaveFocus()
    })
  })

  it('should render with icon on the left', () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<MailOutline data-testid="icon" />}
      />
    )

    const icon = screen.getByTestId('icon')
    expect(icon).toBeInTheDocument()
    expect(icon.parentElement).toHaveStyle({ 'flex-direction': 'row' })
  })

  it('should render with icon on the right', () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        iconPosition="right"
        icon={<MailOutline data-testid="icon" />}
      />
    )

    const icon = screen.getByTestId('icon')
    expect(icon).toBeInTheDocument()
    expect(icon.parentElement).toHaveStyle({ 'flex-direction': 'row-reverse' })
  })

  it('should render a disabled input with correct styles and not change value', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        disabled
      />
    )

    const input = screen.getByLabelText('TextField')
    expect(input).toBeDisabled()
    expect(input).toHaveStyle({ cursor: 'not-allowed' })

    const text = 'Test'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue()
    })
    expect(onInput).not.toHaveBeenCalled()

    expect(screen.getByText('TextField')).toHaveStyle({
      color: '#8F8F8F',
      cursor: 'not-allowed'
    })
  })

  it('shoul not be accessible with tab when disabled', async () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        disabled
      />
    )

    expect(document.body).toHaveFocus()
    userEvent.tab()

    await waitFor(() => {
      expect(screen.getByLabelText('TextField')).not.toHaveFocus()
    })
  })

  it('renders with error', () => {
    const { container } = renderWithTheme(
      <TextField
        icon={<MailOutline data-testid="icon" />}
        label="TextField"
        labelFor="TextField"
        error="Error message"
      />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})
