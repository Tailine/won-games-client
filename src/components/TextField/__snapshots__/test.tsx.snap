// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders with error 1`] = `
.c5 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c4 svg {
  width: 2.2rem;
  fill: #8F8F8F;
}

.c6 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-left: 0.8rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c6::-webkit-input-placeholder {
  color: #8F8F8F;
}

.c6::-moz-placeholder {
  color: #8F8F8F;
}

.c6:-ms-input-placeholder {
  color: #8F8F8F;
}

.c6::placeholder {
  color: #8F8F8F;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
}

.c0 .c3 {
  border: 1px solid #FF6347;
}

.c0 .c1 {
  color: #FF6347;
}

.c0 svg {
  fill: #FF6347;
}

.c7 {
  color: #FF6347;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <svg
      aria-hidden="true"
      class="c5"
      data-testid="icon"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 14H4V8l8 5 8-5v10zm-8-7L4 6h16l-8 5z"
      />
    </svg>
    <input
      class="c6"
      type="text"
      value=""
    />
  </div>
  <p
    class="c7"
  >
    Error message
  </p>
</div>
`;
